src/scheduler.rs

```diff
@@ -146,63 +196,65 @@ impl SchedulerState {
     /// Indicate that the network likely is lost.
     pub(crate) async fn maybe_network_lost(&self, context: &Context) {
         let inner = self.inner.read().await;
-        let stores = match inner.scheduler {
-            Some(ref scheduler) => {
+        let stores = match *inner {
+            InnerSchedulerState::Started(ref scheduler) => {
                 scheduler.maybe_network_lost();
                 scheduler
                     .boxes()
                     .map(|b| b.conn_state.state.connectivity.clone())
                     .collect()
             }
-            None => return,
+            _ => return,
         };
         drop(inner);
         connectivity::maybe_network_lost(context, stores).await;
     }
 
     pub(crate) async fn interrupt_inbox(&self, info: InterruptInfo) {
         let inner = self.inner.read().await;
-        if let Some(ref scheduler) = inner.scheduler {
+        if let InnerSchedulerState::Started(ref scheduler) = *inner {
             scheduler.interrupt_inbox(info);
         }
     }
 
     pub(crate) async fn interrupt_smtp(&self, info: InterruptInfo) {
         let inner = self.inner.read().await;
-        if let Some(ref scheduler) = inner.scheduler {
+        if let InnerSchedulerState::Started(ref scheduler) = *inner {
             scheduler.interrupt_smtp(info);
         }
     }
 
     pub(crate) async fn interrupt_ephemeral_task(&self) {
         let inner = self.inner.read().await;
-        if let Some(ref scheduler) = inner.scheduler {
+        if let InnerSchedulerState::Started(ref scheduler) = *inner {
             scheduler.interrupt_ephemeral_task();
         }
     }
 
     pub(crate) async fn interrupt_location(&self) {
         let inner = self.inner.read().await;
-        if let Some(ref scheduler) = inner.scheduler {
+        if let InnerSchedulerState::Started(ref scheduler) = *inner {
             scheduler.interrupt_location();
         }
     }
 
     pub(crate) async fn interrupt_recently_seen(&self, contact_id: ContactId, timestamp: i64) {
         let inner = self.inner.read().await;
-        if let Some(ref scheduler) = inner.scheduler {
+        if let InnerSchedulerState::Started(ref scheduler) = *inner {
             scheduler.interrupt_recently_seen(contact_id, timestamp);
         }
     }
 }
 
 #[derive(Debug, Default)]
-struct InnerSchedulerState {
-    scheduler: Option<Scheduler>,
-    /// Whether IO should be started if there is no [`IoPausedGuard`] active.
-    started: bool,
-    /// The number of [`IoPausedGuard`]s that are outstanding.
-    paused: u32,
+enum InnerSchedulerState {
+    Started(Scheduler),
+    #[default]
+    Stopped,
```

Comment:
Looks like `Stopped` is not needed, it's just `Paused { started: false, pause_guard_count: 0 }`

Reply:
We could make `pause_guard_count` a `NonZeroU32` so you can't create the `Paused` state with `pause_guard_count=0`.  It avoids this "do we count from 0 or 1" ambiguity.

I'd leave `Stopped` as an explicit state, I think having stopped and paused as distinct cases makes it easier to reason about.

Reply:
Another option is:
```
enum InnerSchedulerState {
    Started(Scheduler),
    Stopped(usize), // the number of pause guards
    Paused(NonZeroUsize)
};
```

Reply:
> Another option is:
> 
> ```
> enum InnerSchedulerState {
>     Started(Scheduler),
>     Stopped(usize), // the number of pause guards
>     Paused(NonZeroUsize)
> };
> ```

Not sure I follow this one.